operators/tokenizer/roberta_tokenizer.cc

```diff
@@ -0,0 +1,251 @@
+#include "bpetokenizer.hpp"
+#include "string_tensor.h"
+#include "unicode.h"
+
+std::list<std::pair<int, int>> offset_map;
+
+// Note: the following logic comes from CPython: unicodetype_db.h (_PyUnicode_IsWhitespace)
+bool IsWithinUnicodeSpace(char32_t ch) {
+  switch (ch) {
+    case 0x0009:
+    case 0x000A:
+    case 0x000B:
+    case 0x000C:
+    case 0x000D:
+    case 0x001C:
+    case 0x001D:
+    case 0x001E:
+    case 0x001F:
+    case 0x0020:
+    case 0x0085:
+    case 0x00A0:
+    case 0x1680:
+    case 0x2000:
+    case 0x2001:
+    case 0x2002:
+    case 0x2003:
+    case 0x2004:
+    case 0x2005:
+    case 0x2006:
+    case 0x2007:
+    case 0x2008:
+    case 0x2009:
+    case 0x200A:
+    case 0x2028:
+    case 0x2029:
+    case 0x202F:
+    case 0x205F:
+    case 0x3000:
+      return true;
+  }
+  return false;
+}
+
+bool IsEmptyuString(const ustring& str) {
+  return std::all_of(str.begin(), str.end(), [](char32_t ch) { return IsWithinUnicodeSpace(ch); });
+}
+
+KernelRobertaBpeTokenizer::KernelRobertaBpeTokenizer(const OrtApi& api, const OrtKernelInfo* info)
+    : BaseKernel(api, info) {
+  std::string vocab = ort_.KernelInfoGetAttribute<std::string>(info, "vocab");
+  if (vocab.empty()) {
+    ORTX_CXX_API_THROW("vocabulary shouldn't be empty.", ORT_INVALID_ARGUMENT);
+  }
+
+  std::string merges = ort_.KernelInfoGetAttribute<std::string>(info, "merges");
+  if (merges.empty()) {
+    ORTX_CXX_API_THROW("merges shouldn't be empty.", ORT_INVALID_ARGUMENT);
+  }
+
+  if (!TryToGetAttribute<int64_t>("padding_length", padding_length_)) {
+    padding_length_ = -1;
+  }
+
+  if (padding_length_ != -1 && padding_length_ <= 0) {
+    ORTX_CXX_API_THROW("padding_length should be more than 0 or equal -1", ORT_INVALID_ARGUMENT);
+  }
+
+  std::stringstream vocabu_stream(vocab);
+  std::stringstream merges_stream(merges);
+  bbpe_tokenizer_ = std::make_shared<VocabData>();
+  bbpe_tokenizer_->Load(vocabu_stream, merges_stream, "<|endoftext|>", "<|endoftext|>");
+}
+
+std::vector<int64_t> KernelRobertaBpeTokenizer::Tokenize(ustring& input, int64_t max_length) {
+  std::vector<int64_t> res;
+
+  if (IsEmptyuString(input)) {
+    return res;
+  }
+  // Add BOS token to result
+  res.push_back(bbpe_tokenizer_->GetEncoding("<s>"));
+
+  // Parse input
+  auto special_token_split_res = bbpe_tokenizer_->SplitBySpecialTokens(input);
+  TokenWithRegularExp regcmp;
+
+  for (auto& seg_id : special_token_split_res) {
+    if (static_cast<int64_t>(res.size()) >= max_length) break;
+
+    if (seg_id.second != -1) {
+      res.push_back(seg_id.second);
+      continue;
+    }
+
+    auto cur_input = std::move(seg_id.first);
+    // Note: keep ptr to make sure the string_view is valid in the following process
+    const char32_t* ptr = cur_input.c_str();
+    regcmp.Set(ptr);
+
+    int offset = 0;
+    std::list<std::pair<int, int>> offset_mapping;
+
+    // Add offset mapping for BOS token
+    offset_mapping.push_back(std::make_pair(0, 0));
```

Comment:
So, offset-mapping is only available for Roberta tokenizer, can we make it be available for other tokenizer as well?

Reply:
Yes, I believe so. I will go ahead and add this for other tokenizers once we have finalized the roberta tokenizer.